File: sync.go
//@author Stanislav Polaniev <spolanyev@gmail.com>

package main

import (
	"bufio"
	"fmt"
	"os"
	"path"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

var startTime = time.Now()
var wordCount = 0
var selectedWords []string

// var directory = filepath.FromSlash("./../data/a")
var directory = filepath.FromSlash("./../../../private/speed-comparison/data/a")

func main() {
	if fullPathDirectory, err := filepath.Abs(directory); err == nil {
		if _, err := os.Stat(fullPathDirectory); err == nil {
			if entries, err := os.ReadDir(fullPathDirectory); err == nil {
				for _, entry := range entries {
					wordCount += 1
					fullPathFrequency := path.Join(fullPathDirectory, entry.Name(), "2015-2017-spoken-frequency.txt")
					if _, err := os.Stat(fullPathFrequency); err == nil {
						if slice, err := os.ReadFile(fullPathFrequency); err == nil {
							if frequency, err := strconv.Atoi(string(slice)); err == nil {
								if frequency >= 798 {
									fullPathTranslation := path.Join(fullPathDirectory, entry.Name(), "translation.txt")
									if _, err := os.Stat(fullPathTranslation); err == nil {
										word := ""
										if fileDescriptor, err := os.Open(fullPathTranslation); err == nil {
											reader := bufio.NewReaderSize(fileDescriptor, 128)
											if word, err = reader.ReadString('\n'); err == nil {
												word = strings.TrimSpace(word)
												if err := fileDescriptor.Close(); err != nil {
													fmt.Println("Error closing file:", err)
												}
											}

											if word == "" {
												word = entry.Name()
											}
											selectedWords = append(selectedWords, word)
										}
									}
								}
							}
						}
					}
				}
			}
		}
	}

	fmt.Printf("selected %d words from %d, took %.3f seconds\n", len(selectedWords), wordCount, time.Since(startTime).Seconds())
	/*
		for i, word := range selectedWords {
			fmt.Printf("%d. %s\n", i+1, word)
		}
	*/
}

File: sync.js
//@author Stanislav Polaniev <spolanyev@gmail.com>

'use strict';

const path = require('node:path');
const fs = require('node:fs');

const startTime = Math.floor(new Date().getTime());
let wordCount = 0;
const selectedWords = [];

const fullPathDirectory = path.join(__dirname, '..', 'data', 'a');
if (fs.statSync(fullPathDirectory).isDirectory()) {
    const directory = fs.opendirSync(fullPathDirectory);
    let entry;
    while (null !== (entry = directory.readSync())) {
        wordCount++;
        const fullPathFrequency = path.join(fullPathDirectory, entry.name, '2015-2017-spoken-frequency.txt');
        if (fs.statSync(fullPathFrequency).isFile()) {
            const number = fs.readFileSync(fullPathFrequency, 'utf8');
            if (parseInt(number) >= 798) {
                const fullPathTranslation = path.join(fullPathDirectory, entry.name, 'translation.txt');
                if (fs.statSync(fullPathTranslation).isFile()) {
                    
                    let word = '';
                    const chunk = Buffer.alloc(128);
                    let bytesRead = 0;
                    let remaining = '';
                    let isNewLineFound = false;

                    const filePointer = fs.openSync(fullPathTranslation, 'r');
                    while ((bytesRead = fs.readSync(filePointer, chunk, 0, chunk.length)) > 0) {
                        remaining += chunk;
                        let newlineIndex;
                        if ((newlineIndex = remaining.indexOf('\n')) !== -1) {
                            isNewLineFound = true;
                            word = remaining.substring(0, newlineIndex).trim();
                            break;
                        }
                    }

                    if (!isNewLineFound) {
                        word = remaining.trim();
                    }

                    if ('' === word) {
                        word = entry.name;
                    }
                    selectedWords.push(word);
                }
            }
        }
    }
    directory.closeSync();
}

console.log('selected ' + selectedWords.length + ' words from ' + wordCount + ', took ' + (Math.floor(new Date().getTime()) - startTime) / 1000 + ' seconds');
/*
selectedWords.forEach((word, i) => {
    console.log('' + ++i + '. ' + word);
});
*/
